src/lib/formats/ap2_dsk.cpp

```diff
@@ -542,37 +544,39 @@ bool a2_16sect_format::save(util::random_read_write &io, const std::vector<uint3
-								// if the current read is good, and postamble is good, write it in, no matter what.
-								// if the current read is good and the current postamble is bad, write it in unless the postamble was good before
-								// if the current read is bad and the current postamble is good and the previous read had neither good, write it in
-								// if the current read isn't good and neither is the postamble but nothing better
-								// has been written before, write it anyway.
-								if ( ((data[0x156] == c) && (dpost&0xFFFF00)==0xDEAA00) ||
-													(((data[0x156] == c) && (dpost&0xFFFF00)!=0xDEAA00) && ((visualgrid[se][track]&DATAPOST)==0)) ||
-													(((data[0x156] != c) && (dpost&0xFFFF00)==0xDEAA00) && (((visualgrid[se][track]&DATAGOOD)==0)&&(visualgrid[se][track]&DATAPOST)==0)) ||
-													(((data[0x156] != c) && (dpost&0xFFFF00)!=0xDEAA00) && (((visualgrid[se][track]&DATAGOOD)==0)&&(visualgrid[se][track]&DATAPOST)==0))
-													) {
+							if constexpr (USE_OLD_BEST_SECTOR_PRIORITY) {
+								if ((visualgrid[se][track]&DATAGOOD)==0) {
 									for(int i=0x56; i<0x156; i++) {
 										uint8_t dv = data[i];
 										*dest++ = dv;
 									}
 								}
 							}
-#else
-							if ((visualgrid[se][track]&DATAGOOD)==0) {
-								for(int i=0x56; i<0x156; i++) {
-									uint8_t dv = data[i];
-									*dest++ = dv;
+							else {
+								if (((visualgrid[se][track]&DATAGOOD)==0)||((visualgrid[se][track]&DATAPOST)==0)) {
+									// if the current read is good, and postamble is good, write it in, no matter what.
+									// if the current read is good and the current postamble is bad, write it in unless the postamble was good before
+									// if the current read is bad and the current postamble is good and the previous read had neither good, write it in
+									// if the current read isn't good and neither is the postamble but nothing better
+									// has been written before, write it anyway.
+									if ( ((data[0x156] == c) && (dpost&0xFFFF00)==0xDEAA00) ||
+														(((data[0x156] == c) && (dpost&0xFFFF00)!=0xDEAA00) && ((visualgrid[se][track]&DATAPOST)==0)) ||
+														(((data[0x156] != c) && (dpost&0xFFFF00)==0xDEAA00) && (((visualgrid[se][track]&DATAGOOD)==0)&&(visualgrid[se][track]&DATAPOST)==0)) ||
+														(((data[0x156] != c) && (dpost&0xFFFF00)!=0xDEAA00) && (((visualgrid[se][track]&DATAGOOD)==0)&&(visualgrid[se][track]&DATAPOST)==0))
+														) {
+										for(int i=0x56; i<0x156; i++) {
+											uint8_t dv = data[i];
+											*dest++ = dv;
+										}
+									}
 								}
 							}
-#endif
 							// do some checking
-							#ifdef VERBOSE_SAVE
-							if ((data[0x156] != c) || (dpost&0xFFFF00)!=0xDEAA00)
-								fprintf(stderr,"Data Mark:\tChecksum xpctd %d found %d: %s, Postamble %03X: %s\n", data[0x156], c, (data[0x156]==c)?"OK":"BAD", dpost, (dpost&0xFFFF00)==0xDEAA00?"OK":"BAD");
-							#endif
+							if constexpr (VERBOSE_SAVE) {
```

Comment:
I don't think it's necessary to prepend these with `constexpr`, as MAME has the appropriate compiler flags set to handle effective `if (0)` and `if (1)` situations without warnings.

Reply:
Sure, but is there any harm in it? I think it's a useful reminder that it's supposed to be a compile-time selection.

Reply:
That’s not what `if constexpr` is for.  The purpose of `if constexpr` is for when one branch is malformed when the expression is false.  For example see use in src.emu/device.h:
```c++
    template <class DeviceClass>
    DeviceClass *subdevice(std::string_view tag) const
    {
        device_t *const found = subdevice(tag);
        if constexpr (std::is_base_of_v<device_t, DeviceClass>)
        {
            return downcast<DeviceClass *>(found);
        }
        else
        {
            auto const result = dynamic_cast<DeviceClass *>(found);
#if defined(MAME_DEBUG)
            if (found && !result)
                report_bad_device_cast(found, typeid(device_t), typeid(DeviceClass));
#endif
            return result;
        }
    }
```

The code can use the `downcast` helper if `DeviceClass` is derived from `device_t`, but this is invalid otherwise (e.g. if it is derived from `device_interface` but not `device_t`).

Or consider this code in src/lib/util/delegate.h where it needs to make a decision on whether it can cast the call operator to a usable function type or it needs to wrap it in a lambda to convert the return value:
```c++
    template <typename T>
    static functoid_setter make_functoid_setter()
    {
        if constexpr (matching_non_const_call(&unwrapped_functoid_t<T>::operator()))
        {
            return
                    [] (delegate &obj)
                    {
                        obj.basetype::operator=(
                                basetype(
                                    static_cast<ReturnType (unwrapped_functoid_t<T>::*)(Params...)>(&unwrapped_functoid_t<T>::operator()),
                                    obj.unwrap_functoid<T>()));
                    };
        }
        else if constexpr (matching_const_call(&unwrapped_functoid_t<T>::operator()))
        {
            return
                    [] (delegate &obj)
                    {
                        obj.basetype::operator=(
                                basetype(
                                    static_cast<ReturnType (unwrapped_functoid_t<T>::*)(Params...) const>(&unwrapped_functoid_t<T>::operator()),
                                    obj.unwrap_functoid<T>()));
                    };
        }
        else
        {
            return
                    [] (delegate &obj)
                    {
                        obj.basetype::operator=(
                                basetype(
                                    [] (unwrapped_functoid_t<T> &f, Params... args) { return ReturnType(f(std::forward<Params>(args)...)); },
                                    obj.unwrap_functoid<T>()));
                    };
        }
    }
```

Depending on the signature of the call operator, one or two of the branches will be malformed.

Reply:
You can find plenty of examples of testing compile-time constant expressions with regular `if`, e.g. in src/devices/cpu/e132xs/e132xsop.hxx:

```c++
template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
void hyperstone_device::hyperstone_and()
{
        check_delay_PC();

        const uint32_t dst_code = DST_GLOBAL ? DST_CODE : ((DST_CODE + GET_FP) & 0x3f);
        const uint32_t sreg = SRC_GLOBAL ? m_core->global_regs[SRC_CODE] : m_core->local_regs[(SRC_CODE + GET_FP) & 0x3f];
        const uint32_t dreg = (DST_GLOBAL ? m_core->global_regs : m_core->local_regs)[dst_code] & sreg;

        if (dreg == 0)
                SR |= Z_MASK;
        else
                SR &= ~Z_MASK;

        if (DST_GLOBAL)
                set_global_register(dst_code, dreg);
        else
                m_core->local_regs[dst_code] = dreg;

        m_core->icount -= m_core->clock_cycles_1;
}
```

Reply:
Okay, I removed it.